app/models/ems_refresh/parsers/azure.rb

```diff
@@ -0,0 +1,214 @@
+        get_locations.each do |location|
+          begin
+            series << @vmm.series(location)
+          rescue RestClient::BadGateway
+            next
+          end
+        end
+        series = series.flatten!.uniq
+        process_collection(series, :flavors) { |s| parse_series(s) }
+      end
+
+      def get_availability_sets
+        a_zones = @asm.list
+        process_collection(a_zones, :availability_zones) { |az| parse_az(az) }
+      end
+
+      def get_vms
+        vms = @vmm.get_vms
+        process_collection(vms, :vms) { |instance| parse_vm(instance) }
+      end
+
+      def process_collection(collection, key)
+        @data[key] ||= []
+
+        return if collection.nil?
+
+        collection.each do |item|
+          uid, new_result = yield(item)
+          @data[key] << new_result
+          @data_index.store_path(key, uid, new_result)
+        end
+      end
+
+      def parse_series(s)
+        name = uid = s['name']
+        new_result = {
+          :type           => "SeriesAzure",
+          :ems_ref        => uid,
+          :name           => name,
+          :cpus           => s['numberOfCores'], # where are the virtual CPUs??
+          :cpu_cores      => s['numberOfCores'],
+          :memory         => s['memoryInMB'].to_f,
+
+          # Extra keys
+          :os_disk_size   => (s['osDiskSizeInMB'] * 1024),
+          :swap_disk_size => s['resourceDiskSizeInMB']
+
+        }
+        return uid, new_result
+      end
+
+      def parse_az(az)
+        name = az["name"]
+        id   = az["id"].downcase
+
+        new_result = {
+          :type    => EmsAzure::AvailabilityZone.name,
+          :ems_ref => id,
+          :name    => name,
+        }
+        return id, new_result
+      end
+
+      def parse_vm(vm)
+        uid               = vm.fetch_path('resourceGroup') + "\\" + vm.fetch_path('name')
+        series_name       = vm.fetch_path('properties', 'hardwareProfile', 'vmSize')
+        az                = vm.fetch_path('properties', 'availabilitySet', 'id')
+        series            = @data_index.fetch_path(:flavors, series_name)
+
+        new_result = {
+          :type             => 'VmAzure',
+          :uid_ems          => uid,
+          :ems_ref          => uid,
+          :name             => vm.fetch_path('name'),
+          :vendor           => "Microsoft",
+          :raw_power_state  => vm["powerStatus"],
+          :operating_system => process_os(vm),
+          :flavor           => series,
+          :location         => uid,
+          :hardware         => {
+            :disks    => [], # Filled in later conditionally on flavor
+            :networks => [], # Filled in later conditionally on what's available
+          },
+        }
+        new_result[:availability_zone] = fetch_az(az) unless az.nil?
+
+        populate_hardware_hash_with_disks(new_result[:hardware][:disks], vm)
+        populate_hardware_hash_with_series(new_result[:hardware], vm, series)
```

Comment:
Oh ok...I think what confused me is that the other two method creates subcollections, but this one changes actual properties.  If the method were renamed to, say, `populate_hardware_hash_with_series_attributes`, it might be clearer, but it's up to you.